### PR TITLE
enforce `gradients` as an "auxiliary" module

### PR DESCRIPTION
**Pull request by andrijapau**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/gradients/hadamard_gradient.py [hunk 1]
```diff
@@ -21,10 +21,17 @@
 
 import numpy as np
 
-import pennylane as qml
-from pennylane import transform
+from pennylane import math
+from pennylane import ops as qops
```
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/gradients/hadamard_gradient.py:25]: Not a standard, but I've found that it's pretty common for "ops" to be used as a variable name and so you would get a clash in the namespace. But maybe you're right - perhaps I'm a bit too loose on the names for the module alias ... 🤔

[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/hadamard_gradient.py:25]: Brilliant, thank you! I don't dislike them, but it's hard to keep those things unified, so we can try early on :)

[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/gradients/hadamard_gradient.py:25]: @dwierichs, after speaking with core it seems like we can just use the module name by default since PyLint will detect any name space clashing (see picture). But if there is existing name space clashing an alias like this is fine.

image

File: pennylane/gradients/pulse_gradient.py [hunk 2]
```diff
@@ -20,10 +20,13 @@
 
 import numpy as np
 
-import pennylane as qml
-from pennylane import transform
+from pennylane import math
+from pennylane.ops import PauliRot, exp
+from pennylane.ops import functions as qfuncs
```
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/gradients/pulse_gradient.py]: See https://github.com/PennyLaneAI/pennylane/pull/7416/commits/02c791bec740c1c311437745fbc1e5b3199a219b

File: pennylane/gradients/hadamard_gradient.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/gradients/hadamard_gradient.py]: See https://github.com/PennyLaneAI/pennylane/pull/7416/commits/72197bbb7a32bece15088ff3fc33c70918dcb7b3

File: pennylane/gradients/parameter_shift_cv.py [hunk 3]
```diff
@@ -487,14 +490,15 @@ def processing_fn(results):
         # In the future, we might want to change this so that only tuples
         # of arrays are returned.
         for i, g in enumerate(grads):
-            g = qml.math.convert_like(g, results[0])
+            g = math.convert_like(g, results[0])
             if hasattr(g, "dtype") and g.dtype is np.dtype("object"):
-                grads[i] = qml.math.hstack(g)
-        return qml.math.T(qml.math.stack(grads))
+                grads[i] = math.hstack(g)
```
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/gradients/parameter_shift_cv.py:494]: Indeed - never covered. See [here](https://app.codecov.io/gh/PennyLaneAI/pennylane/blob/master/pennylane%2Fgradients%2Fparameter_shift_cv.py).

File: pennylane/gradients/parameter_shift.py [hunk 4]
```diff
@@ -799,6 +811,7 @@ def _expand_transform_param_shift(
     classical_cotransform=contract_qjac_with_cjac,
     final_transform=True,
 )
+# pylint: disable=too-many-positional-arguments
```
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/gradients/parameter_shift.py:814]: See https://github.com/PennyLaneAI/pennylane/pull/7416/ commits/0b8b9f0cee34e97f4b04ce7a73d733bb9058766c and https://github.com/PennyLaneAI/pennylane/pull/7416/commits/ac000003003ac4feb990782b8881b6a4814c3a97.

File: pennylane/gradients/spsa_gradient.py [hunk 5]
```diff
@@ -92,6 +93,7 @@ def null_postprocessing(results):
     classical_cotransform=contract_qjac_with_cjac,
     final_transform=True,
 )
+# pylint: disable=too-many-positional-arguments
```
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/gradients/spsa_gradient.py]: See https://github.com/PennyLaneAI/pennylane/pull/7416/commits/0b8b9f0cee34e97f4b04ce7a73d733bb9058766c and https://github.com/PennyLaneAI/pennylane/pull/7416/commits/ac000003003ac4feb990782b8881b6a4814c3a97.

File: pennylane/gradients/pulse_gradient.py [hunk 6]
```diff
@@ -107,20 +110,20 @@ def _split_evol_ops(op, ob, tau):
...
+        insert_ops = [PauliRot(shift, word, ob.wires) for shift in [np.pi / 2, -np.pi / 2]]
         coeffs = [prefactor, -prefactor]
     else:
         with warnings.catch_warnings():
             if len(ob.wires) <= 4:
                 warnings.filterwarnings(
                     "ignore", ".*the eigenvalues will be computed numerically.*"
                 )
-            eigvals = qml.eigvals(ob)
+            eigvals = eigvals(ob)
```
Comment (JerryChen97) [inline pennylane/gradients/pulse_gradient.py:124]: oh s**t now I understand why we needed it

File: pennylane/gradients/pulse_gradient.py [hunk 7]
```diff
@@ -20,10 +20,13 @@
 
 import numpy as np
 
-import pennylane as qml
-from pennylane import transform
+from pennylane import math
+from pennylane.ops import PauliRot, exp
+from pennylane.ops.functions import dot, eigvals
```
Comment (JerryChen97) [inline pennylane/gradients/pulse_gradient.py:25]: why codefactor said unused? We did use it no?

File: pennylane/gradients/pulse_gradient.py [hunk 8]
```diff
@@ -107,20 +110,20 @@ def _split_evol_ops(op, ob, tau):
...
         coeffs = [prefactor, -prefactor]
     else:
         with warnings.catch_warnings():
             if len(ob.wires) <= 4:
                 warnings.filterwarnings(
                     "ignore", ".*the eigenvalues will be computed numerically.*"
                 )
-            eigvals = qml.eigvals(ob)
+            eigvals = eigvals(ob)
         coeffs, shifts = zip(*generate_shift_rule(eigvals_to_frequencies(tuple(eigvals))))
```
Comment (JerryChen97) [inline pennylane/gradients/pulse_gradient.py:125]:
```suggestion
            eigenvalues = eigvals(ob)
        coeffs, shifts = zip(*generate_shift_rule(eigvals_to_frequencies(tuple(eigenvalues))))
```

Comment (andrijapau) [reply pennylane/gradients/pulse_gradient.py]: Resolved locally!

File: pennylane/gradients/pulse_gradient.py (hunk 7 above)
Comment (andrijapau) [reply pennylane/gradients/pulse_gradient.py:25]: Resolved locally.

File: pennylane/gradients/hadamard_gradient.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/hadamard_gradient.py]: Nice, thanks! :+1:

File: pennylane/gradients/parameter_shift.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/parameter_shift.py:814]: Sorry in case I'm being annoying, but could it be that `argnum` is frequently used positionally in user code? It's a very common first keyword argument in differentiation related code, so I'd definitely see someone write
```python
jax.jacobian(my_fun, (1, 2, 3))
```
and then maybe also `spsa_grad(my_qnode, (1, 2, 3))`? 

Could be something to verify with product, but I think it's good to keep the chance of a breaking change effect to a minimum, and one positional keyword argument does not hurt, I suppose :)

File: pennylane/gradients/parameter_shift_cv.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/parameter_shift_cv.py:494]: Oh wow :D

**Review by dwierichs — APPROVED** [review]
Looks good to me, thanks for the changes :+1:
I think I'd vote for allowing `argnum` to be positional keyword argument, but otherwise looks merge-ready to me :)

**Comment by andrijapau** [comment]
CC @dwierichs @JerryChen97 - reverted the forcing of kwargs in this PR as,
1. they are beyond the scope of this PR
2. will be addressed this quarter when we update the PyLint version to latest
3. need input from product

File: pennylane/gradients/parameter_shift.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/gradients/parameter_shift.py:814]: Good point! This will be addressed later in the quarter now.
